### PR TITLE
Handle null byte when serving static files

### PR DESCRIPTION
**Pull request by makushline**
This is raising an exception on a production application I am running. This should be fine since `nil` is an acceptable return type of the method.

### CODE REVIEW

**Comment by jkowens** [comment]
Rack has a utility method `valid_path?` that checks to see if it contains a null byte. If it's not a valid path that could be part of the return condition. Can you also add test for this scenario?

File: lib/sinatra/base.rb [hunk 1]
```diff
@@ -1057,13 +1057,16 @@ def route_missing
     # Attempt to serve static files from public directory. Throws :halt when
     # a matching file is found, returns nil otherwise.
     def static!(options = {})
```
Comment (makushline) [inline lib/sinatra/base.rb]: I thought this was slightly clearer but happy to implement it differently 😀.

**Comment by makushline** [comment]
@jkowens I missed your note and I had snoozed the errors on my application for some time. It's only when they came again that I remember about the open PR.

**Comment by jkowens** [comment]
LGTM, thanks! I'll see if @namusyaka can review 👍

File: lib/sinatra/base.rb [hunk 2]
```diff
@@ -1057,13 +1057,16 @@ def route_missing
...
-      return if (public_dir = settings.public_folder).nil?
-      path = File.expand_path("#{public_dir}#{URI_INSTANCE.unescape(request.path_info)}" )
-      return unless File.file?(path)
-
-      env['sinatra.static_file'] = path
+      public_dir = settings.public_folder
```
Comment (namusyaka) [inline lib/sinatra/base.rb]: Why did you move this up here? I don't see why the one-liner is removed.

File: test/static_test.rb [hunk 3]
```diff
@@ -59,6 +59,12 @@ class StaticTest < Minitest::Test
     assert not_found?
   end
 
+  it 'passes to the next handler when the path contains null bytes' do
+    @app.set :public_folder, "\0"
+    get "/"
+    assert not_found?
```
Comment (namusyaka) [inline test/static_test.rb]: I don't think this test covers the case correctly, because [the `:static` is set to `true`](https://github.com/sinatra/sinatra/blob/a4dd24add24f2dd0e7299b9e68e12038138294d3/test/static_test.rb#L6).
If you don't set `:static` and [use the default settings](https://github.com/sinatra/sinatra/blob/a4dd24add24f2dd0e7299b9e68e12038138294d3/lib/sinatra/base.rb#L1829), `File.exist?` will continue to raise an exception as you said. The `static?` is called in [dispatching](https://github.com/sinatra/sinatra/blob/a4dd24add24f2dd0e7299b9e68e12038138294d3/lib/sinatra/base.rb#L1096).

Comment (jkowens) [reply test/static_test.rb]: Ah right. The request path is what we need to test, not the public folder. Can we change the test to look at a request path with a null byte?

```
get "/foo%00"
assert not_found?
````

I would expect the app to raise an exception if `public_folder` path did contain a null byte.

File: lib/sinatra/base.rb (hunk 2 above)
Comment (makushline) [reply lib/sinatra/base.rb]: This was done for clarity but shouldn't change functionality. I don't feel strongly about it though; if you feel strongly I'm happy to use the one-liner.

File: test/static_test.rb (hunk 3 above)
Comment (makushline) [reply test/static_test.rb]: @namusyaka & @jkowens that makes sense. I've amended the commit to reflect your feedback.

Comment (makushline) [reply test/static_test.rb]: @namusyaka & @jkowens bump for the message above 😁. How does this change look?

File: lib/sinatra/base.rb (hunk 2 above)
Comment (jkowens) [reply lib/sinatra/base.rb]: @makushline why don't we go back to the one-liner for now.

File: lib/sinatra/base.rb [hunk 4]
```diff
@@ -1057,13 +1057,16 @@ def route_missing
...
+      return if public_dir.nil?
+      path = "#{public_dir}#{URI_INSTANCE.unescape(request.path_info)}"
+      return unless valid_path?(path)
+      expanded_path = File.expand_path(path)
```
Comment (jkowens) [inline lib/sinatra/base.rb]: @makushline can you stick with the `path` variable instead of introducing `exanded_path`?

File: test/static_test.rb (hunk 3 above)
Comment (jkowens) [reply test/static_test.rb]: @makushline looks good. I think a couple small tweaks and this should be ready to merge.
